[BASIC] Fix 1E39 bug (Issue #247)

calmsacibis995 opened the pull request:
The following modifications in `basic/code17.s` fixes the 1E39 bug (see #247) that we inherited from the C64.
![2024-08-31_23-46](https://github.com/user-attachments/assets/03e5ddad-268c-4f0f-8145-a8116f45d68d)

mooinglemur commented on basic/code17.s:
```diff
@@ -4,36 +4,45 @@ conint	jsr posint
-val2	stx index2+1
-	lda (index2)
+val2	stx strng2+1
+	lda (strng2),y	;replace character after string with $00 (fake eol)
```
What is the value of Y here?  It's the high byte of txtptr, which makes no sense as an offset to `strng2`.

I don't think this PR solves the issue.  In fact, it breaks `VAL()` when you give it a proper argument.

mooinglemur replied:
![image](https://github.com/user-attachments/assets/79902443-82a1-434b-a54d-598d370e279f)

mooinglemur replied:
![image](https://github.com/user-attachments/assets/b3c5b16b-7648-4138-90ce-f82fc5f87f42)

calmsacibis995 replied:
Intresting, because I used `STA (),Y` and it seemed to fix the issue, but now that you told me that it breaks `VAL()` when given real arguments, I have to re-work my solution.

calmsacibis995 commented:
I replaced `strng2` with `index2`, and it no longer errors out when using `VAL()` with valid parameters, but the comment is still cut off when it exceeds a set number of characters.
![2024-09-02_17-41](https://github.com/user-attachments/assets/9054bc54-ace6-4078-9b3a-8f3812a078c7)

mooinglemur commented:
I don't think any trivial changes like this would solve this complete.  It all comes down to how passing strings around happens in BASIC.

If the string passed to `VAL()` is a string literal, let's take your example. In the original code, the final quote is replaced with a null, and thus the terminated string `1E39` is passed to the math routine.

Your usage of `sta (##),y` simply places the null byte somewhere where it doesn't belong, which is a little bit further into the program source.  The reason this gives the right answer is that there is a null character at the end of the string.  Passing the string `1234")` into the assembly routine gives the right answer, simply because the punctuation at the end of the string does not affect the evaluation of the string into the number.

Where the lack of termination in the correct spot will certainly frequently break is if you call `VAL(A$)`.  Since you're passing in a variable, the value is coming from the string heap.  If you pass in an unterminated string, the assembly routine may not find a null character anywhere within the first 256 bytes after the number starts, or it may find one but the routine gets a combined string that evaluates into a different number.

The most simple solutions here would probably involve allocating a new string off the heap with a length of the original string +1, and then copying the data to the new location while terminating it, and then allowing garbage collection to reclaim the space later.

calmsacibis995 commented:
I reverted my chages from `sta (##),y` to `sta abs` and reworked the function by handling strings properly, allocating a string using the `strspa` function, with a length of the original +1, as you said above.

 As you can see in the screenshot below, the comment is not destroyed, regardless of its length, due to proper string handling, but the `VAL()` error message is reporting the wrong answer when given real use cases.

e.g. line 10: `PRINT VAL("1234")` should report `1234` and exit without an error message. But it reports 0 and throws an error message.

![2024-09-03_13-24](https://github.com/user-attachments/assets/198b282a-c2ad-4cda-9742-7012f45cba30)

mooinglemur commented:
perhaps some cheap bounds checking after `inc` could be useful too (beq to string too long error in case we have a 255 character string)

mooinglemur commented:
As an aside, for code style reasons, please use spaces for comments on the same line as assembly code rather than tabs.  There should be a maximum of one tab per line, and the first indent character should always be a tab, allowing for good visual alignment regardless of tab width.

calmsacibis995 commented:
I went with your suggestion, and it fixed the issue. I made my modifications to handle the pointers correctly at the end, by pulling them from the stack that we had saved previously, and added an `rts` and the end. I also added a check for a string that is 255 characters long (the null byte is included).

Even though there is a `valrts` function that is supposed to end the function, when I removed it, thinking that there was no point in keeping it, it breaks `VAL()`, so I kept it.

![2024-09-04_11-02](https://github.com/user-attachments/assets/556121c4-82cb-4b43-afbd-0f93afba41a5)

calmsacibis995 commented:
I believe now it can be merged, as a proper, permanent fix has been implemented.

mooinglemur commented on basic/code17.s:
```diff
@@ -4,36 +4,62 @@ conint	jsr posint
 	pha
-	lda #0
-	sta (index2)
+	inc
```
below this `inc`, we can add `beq val_strlong`.  This means we went from 255 to 0.  Or 255 to 256, which means the string is too long to terminate.  No need for cmp here as the Z flag is set if `inc` causes the .A register to become 0.

mooinglemur commented on basic/code17.s:
```diff
@@ -4,36 +4,62 @@ conint	jsr posint
+	ply
+	stx txtptr
+	sty txtptr+1
+	plx            ;restore saved index1
+	ply
+	stx index1
+	sty index1+1
```
This takes more off the stack than we put on, and results in a crash.  We can remove this completely as we already restored `index1` above.

mooinglemur commented on basic/code17.s:
```diff
@@ -4,36 +4,62 @@ conint	jsr posint
 	sty txtptr+1
-valrts	rts
+valrts	rts        ;done!
+val_strlong	jmp zerofc
```
This simply returns "0" as the value, which is not what we want.  We want to generate a STRING TOO LONG error and stop execution if the string is indeed too long.  To do this, the simplest way is to
```
ldx #errls 
jmp error
```

mooinglemur commented:
Unfortunately with the code you pushed, this is not what happens for me.
![image](https://github.com/user-attachments/assets/4d49cefb-b87d-4afb-8163-a9a1fbad59b5)

calmsacibis995 commented:
OK, let's do what you suggested. This should fix it.

mooinglemur reviewed (approved):
Thanks!
